server: Catch panics in gRPC for DB console requests

alyshanjahani-crl opened the pull request:
Previously if a gRPC handler had an uncaught panic this would cause the node to crash. This is not desired when the gRPC handler is serving a DB console request, i.e. uncaught panics due to DB console functionality should not cause the CRDB node to crash.

This commit adds metadata to the context in requests that originate from the gRPC gateway. These correspond to requests from the DB console. Now the gRPC server in CRDB can add a panic recovery interceptor and determine whether or not to recover a panic.

Fixes: #140276 
Release note (bug fix): Unexpected errors in DB console no longer crash the node.

alyshanjahani-crl committed:
server: Catch panics in gRPC for DB console requests

Previously if a gRPC handler had an uncaught panic this would cause the node
to crash. This is not desired when the gRPC handler is serving a DB console
request, i.e. uncaught panics due to DB console functionality should not cause
the CRDB node to crash.

This commit adds metadata to the context in requests that originate from the
gRPC gateway. These correspond to requests from the DB console. Now the gRPC
server in CRDB can add a panic recovery interceptor and determine whether or
not to recover a panic.

Fixes: #47076
Release note (bug fix): Unexpected errors in DB console no longer
crash the node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145563)
<!-- Reviewable:end -->

alyshanjahani-crl commented on pkg/server/grpc_gateway.go:
```diff
@@ -69,6 +71,9 @@ func configureGRPCGateway(
 		gwruntime.WithOutgoingHeaderMatcher(authserver.AuthenticationHeaderMatcher),
 		gwruntime.WithMetadata(authserver.TranslateHTTPAuthInfoToGRPCMetadata),
+		gwruntime.WithMetadata(func(ctx context.Context, r *http.Request) metadata.MD {
+			return metadata.Pairs("http-path", r.URL.Path)
```
i don't think i can put this constant in the apiconstants package (which is under server package). Since `server` imports `rpc`.
Would it be fine to put this in `pkg/base/constants.go` ?

cthumuluru-crdb replied:
`pkg/base/constants.go` doesn't looks like the right place for this. Unfortunately I don't have a definitive answer here. Can you take a look at `pkg/rpc/auth.go` and see if something similar can be done here too? Basically move set and get metadata methods to a new file in rpc package?

cthumuluru-crdb replied:
What do you think about moving interceptor into a different file?
<!-- Reviewable comment -OPiwUnxAkHksda5xCr1:b-2zspme -->
<!-- Sent from Reviewable.io -->

alyshanjahani-crl replied:
i think you might be seeing an old revision? The interceptor is in `pkg/rpc/context.go` or did you mean to move to a different file in the rpc package?
<!-- Reviewable comment -OPksYpa1goWDGSYCeaJ:bbg03ru -->
<!-- Sent from Reviewable.io -->

cthumuluru-crdb commented on pkg/rpc/context.go:
```diff
@@ -187,6 +188,34 @@ func NewServerEx(
+			if len(paths) > 0 {
+				defer func() {
+					if p := recover(); p != nil {
+						logcrash.ReportPanic(context.Background(), nil, p, 1 /* depth */)
```
Use the context provided to the function here?

alyshanjahani-crl replied:
oops, yes fixed thanks

cthumuluru-crdb commented on pkg/rpc/context.go:
```diff
@@ -187,6 +188,34 @@ func NewServerEx(
+		resp, err = handler(ctx, req)
+		return resp, err
+	}
+	unaryInterceptor = append(unaryInterceptor, recoveryInterceptor)
```
The order of inteceptors is important here if we wish to recover from panics caused by other interceptors before the request is handled.

alyshanjahani-crl replied:
Ack, moved to register the interceptor just after the metrics interceptor (which needs to be second as noted in comments)

alyshanjahani-crl commented:
> see if something similar can be done here too? Basically move set and get metadata methods to a new file in rpc package?

Yes thanks for that pointer, i refactored the code so that it's in the rpc package, and that it's purpose is more clear, also added a test.

PTAL

cthumuluru-crdb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145563)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl)



<!-- Sent from Reviewable.io -->

alyshanjahani-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145563)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb)



<!-- Sent from Reviewable.io -->

alyshanjahani-crl commented:
Discussed offline w @cthumuluru-crdb 

Moved the interceptor to `rpc/metrics.go` to keep it coupled w/ other observability related rpc instrumentation.

cthumuluru-crdb commented on pkg/rpc/metrics.go:
```diff
@@ -403,3 +411,33 @@ func NewRequestMetricsInterceptor(
+				if p := recover(); p != nil {
+					logcrash.ReportPanic(ctx, nil, p, 1 /* depth */)
+					// The gRPC gateway will put this message in the HTTP response to the client.
+					err = errors.Newf("an unexpected error occurred: %v", p)
```
Is this message meant to be sent to the client?

alyshanjahani-crl replied:
Oops, nope we want to just say "an unexpected error occurred", and log the panic.

Thanks for catching that.

cthumuluru-crdb reviewed (approved):
Changes look good to me except for one comment. Do we intend to send the panic message to the clients?

alyshanjahani-crl commented:
bors r+
